### PR TITLE
Add Canadian French input locale

### PR DESCRIPTION
**Pull request by Nebuleon**
## Description

This pull request introduces the Canadian French input locale, otherwise known as French Canadian or French (Canada), to QMK. This is not the input locale called "Canadian French (Legacy)".

The commit includes declarations in the required .hjson file, the keymap header and the file for use with `SEND_STRING`, as well as the reference to those files in the documentation. Key names start with `FC_` for French Canadian; as for their 4-character names, I have reused the names of keys that output the respective characters in any other existing input locale; and I have validated the key labels, keyboard diagrams, keycodes, Shift and AltGr use, as well as the lookup tables for the `SEND_STRING` macro. The input locale works on both Linux and Windows using both layers that output each individual key as well as a `SEND_STRING` that sends the entire ASCII table.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation
- [x] Other: Input locale (`quantum/keymap_extras`)

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Nebuleon** [comment]
GitHub Actions pointed out that the generated file had one extra newline compared to what is expected, so I removed it in a second commit, and now the `regen` check succeeds (and is no longer showing up in the list). This should now be good for review.

File: docs/reference_keymap_extras.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ These headers are located in [`quantum/keymap_extras/`](https://github.com/qmk/q
 
 |Layout                           |Keycodes Header                  |Sendstring LUT Header               |
 |---------------------------------|---------------------------------|------------------------------------|
+|Canadian French                  |`keymap_canadian_french.h`       |`sendstring_canadian_french.h`      |
```
Comment (fauxpark) [inline docs/reference_keymap_extras.md:23]: I think I would name this `French (Canada)` to match the other entries where the convention is `Language (Region)`. Then it should be moved down beneath the Belgian French entry.

File: quantum/keymap_extras/keymap_canadian_french.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    Y8b d88P
+     "Y88P"
+*******************************************************************************/
+
+#pragma once
+#include "keycodes.h"
+// clang-format off
+
+// Aliases
+#define FC_HASH KC_GRV  // #
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_canadian_french.h:31]: Prefix should be `FR`; generally it should match [ISO-3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2).

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson [hunk 3]
```diff
@@ -0,0 +1,407 @@
+{
+    "aliases": {
+/*
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┐
+ * │ # │ 1 │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 0 │ - │ = │       │
```
Comment (fauxpark) [inline data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:5]: This does not quite match the French Canadian layout as the Windows OSK reports it:
image

File: data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        }
+        "S(FC_COMM)": {
+            "key": "FC_QUOT",
+            "label": "'",
+        }
+/* AltGr symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┐
+ * │ \ │ ± │ @ │ £ │ ¢ │ ¤ │ ¬ │ ¦ │ ² │ ³ │ ¼ │ ½ │ ¾ │       │
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┤
+ * │     │   │   │ € │   │   │   │   │   │ § │ ¶ │ [ │ ] │     │
```
Comment (fauxpark) [inline data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:293]: Windows OSK doesn't have a Euro symbol here.

Comment (sigprof) [reply data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:293]: Apparently the Euro symbol is a somewhat recent addition to the Canadian French layout — it appears in [Windows 10.0.22598.1](http://kbdlayout.info/kbdca/10.0.22598.1) (which is apparently a preview release of Windows 11 23H2, released in the Canary channel on 7 April 2023), but does not appear in [Windows 10.0.17134.1](http://kbdlayout.info/kbdca/10.0.17134.1) (Windows 10 1803).  No idea whether that symbol appears there in any proper releases of Windows 10 and Windows 11 — kbdlayout.info does not have any intermediate versions between those two.

File: data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson (hunk 3 above)
Comment (Nebuleon) [reply data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:5]: Which input locale did you use on Windows when reviewing this PR?

Here is a screenshot of what exact input locale this PR refers to:
Canadian French input locale selection

And a screenshot of the "Canadian French" input locale chosen above in Windows 10's On-Screen Keyboard application:
Canadian French input locale as shown on the On-Screen Keyboard, Windows 10

File: data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson (hunk 4 above)
Comment (Nebuleon) [reply data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:293]: The Euro symbol is indeed not output by AltGr+E on Windows 10 22H2 19045.3155, but KDE has had it for a while.

File: docs/reference_keymap_extras.md (hunk 1 above)
Comment (Nebuleon) [reply docs/reference_keymap_extras.md]: Applied in 12815db88f.

File: quantum/keymap_extras/keymap_canadian_french.h (hunk 2 above)
Comment (Nebuleon) [reply quantum/keymap_extras/keymap_canadian_french.h]: Applied in 12815db88f.

File: data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson (hunk 4 above)
Comment (Nebuleon) [reply data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:293]: Is it policy to restrict input locale declarations in QMK to only what the most common released versions of Windows can do? If so, I can remove `FR_EURO` from the list.

**Comment by Nebuleon** [comment]
And another commit because I realised that, while I had checked that everything that *was* in the keymap was emitting the correct keycodes, I had not checked whether everything that *should* have been there *was* indeed there. I missed `FR_RABK`, also known as `>`, and added it to 423b6bfd83.

File: quantum/keymap_extras/sendstring_canadian_french.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Sendstring lookup tables for Canadian French layouts
+
+#pragma once
+
+#include "keymap_canadian_french.h"
+#include "quantum.h"
```
Comment (fauxpark) [inline quantum/keymap_extras/sendstring_canadian_french.h:22]:
```suggestion
#include "send_string.h"
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by Nebuleon** [comment]
Hi there.

This pull request is still relevant, and I have one outstanding question:
> Is it policy to restrict input locale declarations in QMK to only what the most common released versions of Windows can do? If so, I can remove `FR_EURO` from the list.

**Comment by fauxpark** [comment]
Basically, yes, because it's better to have keycodes "omitted" from the headers compared to Windows 11, than to have users complain about the keycodes not working on older versions. Perhaps this could be mitigated with notes in the docs, but I have observed many times already users not seeming to even notice or understand the last three columns of [this table](https://docs.qmk.fm/#/keycodes?id=basic-keycodes). So I have little faith it will do much.

**Comment by Nebuleon** [comment]
Right. I have deleted `FR_EURO` from the data to ease that burden from user complaints. It's still easy for a user to use `ALGR(FR_E)` if they need the euro symbol in Canadian French.

I do notice that the PR Regenerate Files test fails, after using this command:
`qmk generate-keycode-extras -l canadian_french -v 0.0.1 > quantum/keymap_extras/keymap_canadian_french.h`
I'm not sure why PR Regenerate Files wants the final newline removed, but I will delete that too,

File: data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+        }
+/* AltGr symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┐
+ * │ \ │ ± │ @ │ £ │ ¢ │ ¤ │ ¬ │ ¦ │ ² │ ³ │ ¼ │ ½ │ ¾ │       │
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┤
+ * │     │   │   │   │   │   │   │   │   │ § │ ¶ │ [ │ ] │     │
+ * ├─────┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┐    │
+ * │      │   │   │   │   │   │   │   │   │   │ ~ │ { │ } │    │
+ * ├────┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴───┴────┤
+ * │    │ ° │   │   │   │   │   │   │ µ │ ¯ │shy│ ´ │          │
```
Comment (fauxpark) [inline data/constants/keycodes/extras/keycodes_canadian_french_0.0.1.hjson:301]:
```suggestion
 * │    │ ° │   │   │   │   │   │   │ µ │ ¯ │ - │ ´ │          │
```

File: quantum/keymap_extras/sendstring_canadian_french.h [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    // BS    TAB      LF       VT       FF       CR       SO       SI
+    KC_BSPC, KC_TAB,  KC_ENT,  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    // DLE   DC1      DC2      DC3      DC4      NAK      SYN      ETB
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    // CAN   EM       SUB      ESC      FS       GS       RS       US
+    XXXXXXX, XXXXXXX, XXXXXXX, KC_ESC,  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+
+    //       !        "        #        $        %        &        '
+    KC_SPC,  FC_1,    FC_2,    FC_HASH, FC_4,    FC_5,    FC_7,    FC_COMM,
```
Comment (fauxpark) [inline quantum/keymap_extras/sendstring_canadian_french.h:97]: These should all be `FR_`.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/keymap_extras/sendstring_canadian_french.h (hunk 7 above)
Comment (Nebuleon) [reply quantum/keymap_extras/sendstring_canadian_french.h]: Applied in d328803e740b39deea09d27c3c97c7c4bc757caa.

**Review by fauxpark — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**